### PR TITLE
fix: detect cycles in globals

### LINKED ISSUE #6780: Comptime cyclic globals stack overflow
### Aim

Attempted to compile a program with comptime cyclic globals:

```noir
comptime global A: u32 = B;
comptime global B: u32 = A;

fn main() { }
```

### Expected Behavior

Compilation to fail with a user error

### Bug

Compilation results in a stack overflow:

```bash
❯ ~/.nargo/bin/nargo compile

thread '<unknown>' has overflowed its stack
fatal runtime error: stack overflow
[1]    257937 IOT instruction (core dumped)  ~/.nargo/bin/nargo compile
```

### To Reproduce

1.
2.
3.
4.

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Project Impact

None

### Blocker Context

_No response_

### Nargo Version

nargo version = 0.39.0 noirc version = 0.39.0+df8f2eee5c27d3cd4b6128056afdd9bd4a0322fe (git version hash: df8f2eee5c27d3cd4b6128056afdd9bd4a0322fe, is dirty: false)

### NoirJS Version

_No response_

### Proving Backend Tooling & Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #6780

## Summary

This program:

```noir
comptime global A: u32 = B;
comptime global B: u32 = A;

fn main() { }
```

now gives these errors:

```
error: Variable not in scope
  ┌─ src/main.nr:2:26
  │
2 │ comptime global B: u32 = A;
  │                          - Could not find variable
  │

error: This global recursively depends on itself
  ┌─ src/main.nr:1:26
  │
1 │ comptime global A: u32 = B;
  │                          -
  │

error: Dependency cycle found
  ┌─ src/main.nr:1:17
  │
1 │ comptime global A: u32 = B;
  │                 - 'A' recursively depends on itself: A -> B -> A
  │

Aborting due to 3 previous errors
```

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/tests.rs [hunk 1]
```diff
@@ -3855,3 +3855,23 @@ fn disallows_underscore_on_right_hand_side() {
...
+    "#;
+    let errors = get_program_errors(src);
+
+    assert!(errors.iter().any(|(error, _)| matches!(
+        error,
+        CompilationError::InterpreterError(InterpreterError::GlobalsDependencyCycle { .. })
+    )));
+    assert!(errors.iter().any(|(error, _)| matches!(
+        error,
+        CompilationError::ResolverError(ResolverError::DependencyCycle { .. })
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/tests.rs:3875]: Do we think the two errors may be confusing?

**Review by vezenovm — APPROVED** [review]
One thought, but LGTM

**Review by jfecher — APPROVED** [review]
LGTM. I don't see a way to avoid the double error, and double errors are already common when running comptime code. To avoid them we'd have to return something like `Result<_, Option<Error>>` which would get messy.

**Comment by asterite** [comment]
I was thinking of another way to do this, which is to do it similarly to `FunctionBody`:

```rust
pub enum FunctionBody {
    Unresolved(FunctionKind, BlockExpression, Span),
    Resolving,
    Resolved,
}
```

So a global's value would be a similar enum, and we wouldn't have to add a field to the interpreter. But I don't know which way is preferred...

**Comment by jfecher** [comment]
Either is fine, and either way you'd have to issue an extra error still

**Pull request merged**